lib/remote_io.rb

```diff
@@ -73,63 +79,89 @@ def read(n_bytes)
 
   protected
 
+  REDIRECT_LIMIT = 3
+  UNSAFE_URI_CHARS = %r{[^\-_.!~*'()a-zA-Z\d;/?:@&=+$,\[\]%]}
+
+  # Remove the fragment and escape any unsafe characters in the given URI,
+  # retaining the prior host and scheme if the URI is relative.
+  #
+  # @param uri[String, URI::Generic] The URI to be made safe.
+  def uri=(uri)
+    safe_uri_string = uri.to_s.gsub(UNSAFE_URI_CHARS) do |unsafe_char|
+      "%#{unsafe_char.unpack('H2' * unsafe_char.bytesize).join('%').upcase}"
+    end
+    if safe_uri_string.start_with?('/')
+      @uri.path = safe_uri_string
+    else
+      @uri = URI(safe_uri_string)
+    end
+  end
+
   # Only used internally when reading the remote file
   #
-  # @param range[Range] the HTTP range of data to fetch from remote
-  # @return [String] the response body of the ranged request
-  def request_range(range)
+  # @param range[Range] The HTTP range of data to fetch from remote
+  # @param redirects[Integer] The amount of remaining permitted redirects
+  # @return [[Integer, String]] The response body of the ranged request
+  def request_range(range, redirects = REDIRECT_LIMIT)
     # We use a GET and not a HEAD request followed by a GET because
     # S3 does not allow HEAD requests if you only presigned your URL for GETs, so we
     # combine the first GET of a segment and retrieving the size of the resource
-    conn = Faraday.new(headers: @headers) do |faraday|
-      faraday.use FaradayMiddleware::FollowRedirects
-      # we still need the default adapter, more details: https://blog.thecodewhisperer.com/permalink/losing-time-to-faraday
-      faraday.adapter Faraday.default_adapter
-    end
-    response = conn.get(@uri, nil, range: 'bytes=%d-%d' % [range.begin, range.end])
-
-    case response.status
+    response = Net::HTTP.start(@uri.hostname, @uri.port, use_ssl: @uri.scheme == 'https') { |http|
+      http.request_get(@uri, @headers.merge({ 'range' => 'bytes=%d-%d' % [range.begin, range.end] }))
+    }
+    status = Integer(response.code)
```

Comment:
don't need to convert the status code into an integer. You can use directly `Net::HTTPSuccess` built in

Reply:
and other codes too for the code below

Reply:
How's that looking to you now?